test/common/stats/stats_matcher_test.cc

```diff
@@ -0,0 +1,246 @@
+#include "envoy/config/metrics/v2/stats.pb.h"
+
+#include "common/stats/stats_matcher_impl.h"
+
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+using testing::IsFalse;
+using testing::IsTrue;
+
+namespace Envoy {
+namespace Stats {
+
+class StatsMatcherTest : public testing::Test {
+public:
+  StatsMatcherTest() {}
+  envoy::type::matcher::StringMatcher* inclusionList() {
+    return stats_config_.mutable_stats_matcher()->mutable_inclusion_list()->add_patterns();
+  }
+  envoy::type::matcher::StringMatcher* exclusionList() {
+    return stats_config_.mutable_stats_matcher()->mutable_exclusion_list()->add_patterns();
+  }
+  void rejectAll(const bool should_reject) {
+    stats_config_.mutable_stats_matcher()->set_reject_all(should_reject);
+  }
+
+  void initMatcher() { stats_matcher_impl_ = std::make_unique<StatsMatcherImpl>(stats_config_); }
+
+  envoy::config::metrics::v2::StatsConfig stats_config_;
+
+protected:
+  void expectAccepted(std::vector<std::string> expected_to_pass) {
+    for (const auto& stat_name : expected_to_pass) {
+      EXPECT_THAT(stats_matcher_impl_->rejects(stat_name), IsFalse());
+    }
+  }
+  void expectDenied(std::vector<std::string> expected_to_fail) {
+    for (const auto& stat_name : expected_to_fail) {
+      EXPECT_THAT(stats_matcher_impl_->rejects(stat_name), IsTrue());
+    }
+  }
+
+private:
+  std::unique_ptr<StatsMatcherImpl> stats_matcher_impl_;
+};
+
+TEST_F(StatsMatcherTest, CheckDefault) {
+  // With no set fields, everything should be allowed through.
+  initMatcher();
+  expectAccepted({"foo", "bar", "foo.bar", "foo.bar.baz", "foobarbaz"});
```

Comment:
this pattern is easy to read, but will make any test failures difficult to diagnose, because we won't know which of the strings failed.  This is why the simplest and most common approach is to just write:
`  EXPECT_FALSE(stats_matcher_impl_->rejects("foo"));`
and then a failure will point you to the line of the test that failed. Or you could also write:
`  EXPECT_FALSE(stats_matcher_impl_->rejects("foo")) << "foo";`
which provides you also with the failing string (if it ever fails during dev).

Reply:
I'll use `<<` inside the definitions of expectAccepted() and expectDenied(), so that a failing test prints the string which triggered it.